core/wasm-interface/src/lib.rs

```diff
@@ -0,0 +1,285 @@
+		ptr.ptr as _
+	}
+}
+
+impl<T> IntoValue for Pointer<T> {
+	const VALUE_TYPE: ValueType = ValueType::I32;
+	fn into_value(self) -> Value { Value::I32(self.ptr as _) }
+}
+
+impl<T> TryFromValue for Pointer<T> {
+	fn try_from_value(val: Value) -> Option<Self> {
+		match val {
+			Value::I32(val) => Some(Self::new(val as _)),
+			_ => None,
+		}
+	}
+}
+
+/// The word size used in wasm. Normally known as `usize` in Rust.
+pub type WordSize = u32;
+
+/// The Signature of a function
+#[derive(Eq, PartialEq, Debug, Clone)]
+pub struct Signature {
+	/// The arguments of a function.
+	pub args: Cow<'static, [ValueType]>,
+	/// The optional return value of a function.
+	pub return_value: Option<ValueType>,
+}
+
+impl Signature {
+	/// Create a new instance of `Signature`.
+	pub fn new<T: Into<Cow<'static, [ValueType]>>>(args: T, return_value: Option<ValueType>) -> Self {
+		Self {
+			args: args.into(),
+			return_value,
+		}
+	}
+
+	/// Create a new instance of `Signature` with the given `args` and without any return value.
+	pub fn new_with_args<T: Into<Cow<'static, [ValueType]>>>(args: T) -> Self {
+		Self {
+			args: args.into(),
+			return_value: None,
+		}
+	}
+
+}
+
+/// Something that provides a function implementation on the host for a wasm function.
+pub trait Function {
+	/// Returns the name of this function.
+	fn name(&self) -> &str;
+	/// Returns the signature of this function.
+	fn signature(&self) -> Signature;
+	/// Execute this function with the given arguments.
+	fn execute(
+		&self,
+		context: &mut dyn FunctionContext,
+		args: &mut dyn Iterator<Item=Value>,
+	) -> Result<Option<Value>, String>;
+}
+
+/// Context used by `Function` to interact with the allocator and the memory of the wasm instance.
+pub trait FunctionContext {
+	/// Read memory from `address` into a vector.
+	fn read_memory(&self, address: Pointer<u8>, size: WordSize) -> Result<Vec<u8>, String> {
+		let mut vec = Vec::with_capacity(size as usize);
+		unsafe { vec.set_len(size as usize); }
+		self.read_memory_into(address, &mut vec)?;
+		Ok(vec)
+	}
+	/// Read memory into the given `dest` buffer from `address`.
+	fn read_memory_into(&self, address: Pointer<u8>, dest: &mut [u8]) -> Result<(), String>;
+	/// Write the given data at `address` into the memory.
+	fn write_memory(&mut self, address: Pointer<u8>, data: &[u8]) -> Result<(), String>;
+	/// Allocate a memory instance of `size` bytes.
+	fn allocate_memory(&mut self, size: WordSize) -> Result<Pointer<u8>, String>;
+	/// Deallocate a given memory instance.
+	fn deallocate_memory(&mut self, ptr: Pointer<u8>) -> Result<(), String>;
+	/// Provides access to the sandbox.
+	fn sandbox(&mut self) -> &mut dyn Sandbox;
+}
+
+/// Sandbox memory identifier.
+pub type MemoryId = u32;
+
+/// Something that provides access to the sandbox.
+pub trait Sandbox {
+	/// Get sandbox memory from the `memory_id` instance at `offset` into the given buffer.
+	fn memory_get(
+		&self,
+		memory_id: MemoryId,
+		offset: WordSize,
```

Comment:
I think this should be actually `Pointer` if followed the logic of the other usages of it

Reply:
Isn't that just the offset into the data? Offsets are never given as pointers.

Reply:
well, all wasm pointers are offset from the start of a linear memory : ) This is one of those.